fish-rust/src/locale.rs

```diff
@@ -0,0 +1,133 @@
+/// Support for the "current locale."
+pub use printf_compat::locale::{Locale, C_LOCALE};
+use std::sync::Mutex;
+
+/// Rust libc does not provide LC_GLOBAL_LOCALE, but it appears to be -1 everywhere.
+const LC_GLOBAL_LOCALE: libc::locale_t = (-1_isize) as libc::locale_t;
+
+/// It's CHAR_MAX.
+const CHAR_MAX: libc::c_char = libc::c_char::max_value();
+
+/// \return the first character of a C string, or None if null, empty, has a length more than 1, or negative.
+unsafe fn first_char(s: *const libc::c_char) -> Option<char> {
+    #[allow(unused_comparisons, clippy::absurd_extreme_comparisons)]
+    if !s.is_null() && *s > 0 && *s <= 127 && *s.offset(1) == 0 {
+        Some((*s as u8) as char)
+    } else {
+        None
+    }
+}
+
+/// Convert a libc lconv to a Locale.
+unsafe fn lconv_to_locale(lconv: &libc::lconv) -> Locale {
+    let decimal_point = first_char(lconv.decimal_point).unwrap_or('.');
+    let thousands_sep = first_char(lconv.thousands_sep);
+    let empty = &[0 as libc::c_char];
+
+    // Up to 4 groups.
+    // group_cursor is terminated by either a 0 or CHAR_MAX.
+    let mut group_cursor = lconv.grouping as *const libc::c_char;
+    if group_cursor.is_null() {
+        group_cursor = empty.as_ptr();
+    }
+
+    let mut grouping = [0; 4];
+    let mut last_group: u8 = 0;
+    let mut group_repeat = false;
+    for group in grouping.iter_mut() {
+        let gc = *group_cursor;
+        if gc == 0 {
+            // Preserve last_group, do not advance cursor.
+            group_repeat = true;
+        } else if gc == CHAR_MAX {
+            // Remaining groups are 0, do not advance cursor.
+            last_group = 0;
+            group_repeat = false;
+        } else {
+            // Record last group, advance cursor.
+            last_group = gc as u8;
+            group_cursor = group_cursor.offset(1);
+        }
+        *group = last_group;
+    }
+    Locale {
+        decimal_point,
+        thousands_sep,
+        grouping,
+        group_repeat,
+    }
+}
+
+/// Read the numeric locale, or None on any failure.
+// TODO: figure out precisely which platforms have localeconv_l, or use a build script.
+#[cfg(any(
+    target_os = "macos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+    target_os = "openbsd",
+    target_os = "netbsd",
+))]
```

Comment:
`localeconv_l(3)` is available on BSD, but it doesn't seem to be exposed via the libc crate. Blergh.

Reply:
I think we can work around this by declaring it ourselves in an `extern "C"` block.

Reply:
that seems to have fixed compilation, now there's only a test failure in locale-specific decimal separators on BSD

    3,45: value not completely converted (can't convert ',45')

Reply:
I succcessfully fixed this, it was a real bug.